### ISSUE TITLE
Ensure our tests work with RepeatTests applied to a test class

### ISSUE THREAD

**Issue by Azquelt** [OpenLiberty/open-liberty#2407]
We have a test bucket where we only want to repeat a small selection of tests, so we want to apply the RepeatTests rule at the class level.

Unfortunately, `RepeatTests` currently works by altering all the .xml files it can find in the publish directory. When you request a `LibertyServer` from `LibertyServerFactory` these files are copied from the publish directory to the liberty install directory.

However, once that copy has been done, the files in the publish directory is not used again until the end of the test class when the `FATRunner` cleans up the server. This means that if you apply `RepeatTests` to a class, it does alter the files in `publish` but they then aren't used because they aren't copied to the liberty install directory.

**Comment by aguibert** [comment]
as discussed on slack, we believe this will be resolved by PR #2135

**Comment by Azquelt** [comment]
I'll use this issue to ensure that our tests which use `RepeatTests` on a class are also using `withServer()`.

**Comment by Azquelt** [comment]
This is proving to be more of a pain than expected. The way the logic's written, if `RepeatTests` re-deploys the server while it's running it deletes the whole server directory and you end up with a server you can't shut down or read the logs.

**Comment by aguibert** [comment]
@Azquelt is your test using SharedServer or something that leaves the server running beyond the scope of the test class?  Normally if you just use LibertyServer the server lifecycle only lasts for the scope of a test class, and if you "leak" a running server it will be automatically shut down after the `@AfterClass` runs but before the RepeatTests server.xml replacement happens.

**Comment by Azquelt** [comment]
@aguibert Yeah, we had some tests which store the server object in the suite and all use the same server.

What is stopping the servers between `@AfterClass` and the the server.xml replacement? I couldn't find anything doing that.

I tripped over the following things:
1. (the initial problem) if you're using a `FeatureReplacementAction` on a class, you must supply the servers to replace. If you don't, you don't get an error but it doesn't do what you want.
2. If the `LibertyServer` instance is shared between tests, you must make sure it is stopped before running `FeatureReplacementAction`, otherwise you end up with a zombie server.
3. (slightly tangential) `ShrinkHelper.exportToServer()` writes the application to the server dir both in the Liberty installation, and in the `publish` directory. This means that if two classes use the same test server, when the second one runs, the app from the first test will still be installed. Using `LibertyServer` to delete the app from the server does not remove it from the publish directory.

**Comment by Azquelt** [comment]
Tried commenting out my `@AfterClass` method stop the server and that definitely results in zombie servers and tests that can't read the log file. Our tests are using `SharedServer` so that may be throwing things off but it looked like `FATRunner` had worked around that.

**Comment by aguibert** [comment]
> What is stopping the servers between @AfterClass and the the server.xml replacement? I couldn't find anything doing that.

The bit of code I had in mind was `componenttest.custom.junit.runner.FATRunner.classBlock(RunNotifier)` which calls `LibertyServerFactory.recoverAllServers(getTestClassNameForAssociatedServers());` after the class block finishes running.  However, I just checked FAT trace and see that this runs once after the entire class, and after the repeat actions, which means I was incorrect earlier.

> (the initial problem) if you're using a FeatureReplacementAction on a class, you must supply the servers to replace. If you don't, you don't get an error but it doesn't do what you want.

I believe what's going on here is a quirk in the way `LibertyServer` instances are obtained.  When FATSuite is defined at the class level the `@Server` injection runs twice (the same would occur with a static variable).  When the repeat is defined on the class, the injection / static initializer only runs once.  @olendvcook attempted to fix this by forcing new config to be pushed in the FeatureReplacementAction setup:
```java
        // Make sure config update is pushed to the wlp version of the server & client
        if (serverConfigs.size() > 0) {
            for (String serverName : servers) {
                if (!serverName.equals(ALL_SERVERS))
                    LibertyServerFactory.getLibertyServer(serverName);
            }
        }
```
But that must have some unintended strings attached to it which are causing the zombie servers.

> If the LibertyServer instance is shared between tests, you must make sure it is stopped before running FeatureReplacementAction, otherwise you end up with a zombie server.

This is what I would expect here.  I think hot-swapping feature levels on our normal testing path is non-ideal.

**Comment by Azquelt** [comment]
> This is what I would expect here. I think hot-swapping feature levels on our normal testing path is non-ideal.

I agree. Would it make sense to have `FeatureReplacementAction` stop the server if it's running before doing the feature replacement? This would require passing the `LibertyServer` object to `forServers()` rather than the server name as a string, which might in turn require some extra work to make it work with `@Server` injection.

>  When the repeat is defined on the class, the injection / static initializer only runs once. @olendvcook attempted to fix this by forcing new config to be pushed in the FeatureReplacementAction setup.

Right, and for that code to run, you need to define which servers it should replace. Given that, should we make it throw an error if the rule is attached to a class (rather than a suite) and no servers are defined? I think this would require passing the JUnit test `Description` object through to the `ReplacementAction` somewhere (it has an `isSuite()` method).

Last point: I really don't like that the behaviour of `ShrinkHelper.exportToServer()` is so different to `LibertyServer.copyApplicationToServer()`. Copying the application to the `publish` dir seems completely undesirable. I assume it was done to allow people to set up applications in the FATSuite?

Going forwards, I think that in a lot of cases it makes sense for test classes to build and deploy their own apps (since building an app is straightforwards with Shrinkwrap) rather than relying on one uber-app that all the tests use. This is especially true when you're validating that deployment errors are produced correctly. In this case it doesn't make sense for the app to be copied to the `publish` directory because next time someone calls `getLibertyServer()` (e.g. because you're using `RepeatTests` so your tests get run twice), you end up with *all* your test apps deployed, rather than just the one you're trying to test.

Ideally I'd prefer `exportToServer()` to just export to the server and not to `publish` but I suspect it's too late to change the behaviour of it now. I think a method to just export to the server is necessary (most of the faulttolerance tests should be using it rather than the existing `exportToServer()`) but I'm not sure what it should be called. Any ideas on that one?

**Comment by Azquelt** [comment]
Sudden thought: maybe we could have
```
ShrinkHelper.export(appArchive, libertyServer, APPS);
ShrinkHelper.export(appArchive, libertyServer, DROPINS);
ShrinkHelper.export(appArchive, libertyServer, PUBLISH_APPS_DIR);
ShrinkHelper.export(appArchive, libertyServer, PUBLISH_DROPINS_DIR);
ShrinkHelper.export(libArchive, libertyServer, SERVER_ROOT, "libs");
```

**Comment by aguibert** [comment]
> Would it make sense to have FeatureReplacementAction stop the server if it's running before doing the feature replacement? This would require passing the LibertyServer object to forServers() rather than the server name as a string, which might in turn require some extra work to make it work with @Server injection.

I would like to keep `forServers()` accepting primitive types, since all of this is usually run at static init time.  Additionally, this wouldn't work with `@Server` injection which is undesireable and would confuse most.  Originally my goal was to only have FeatureReplacementAction not interact with LibertyServer objects and only work off of the file system, but that goal has already been broken.  I think a good compromise would be passing in String names of servers, and then looking up the servers from within the FeatureReplacementAction (which Olen was doing with his set of changes).  I think Olen had the right approach, but it just needs to be tweaked a bit.

> Right, and for that code to run, you need to define which servers it should replace. Given that, should we make it throw an error if the rule is attached to a class (rather than a suite) and no servers are defined?

LibertyServer instances should be re-usable even if the config gets swapped around a bit, so I don't think it's necessary to blow up if we can't re-inject/re-init the LibertyServer objectys.  We just need to figure out a way to indicate that the config needs to be pushed from published into the wlp install again (normally it happens on the first startServer() call).  I'd like the behavior of "no servers defined" to remain defaulted to updating all servers. 

> Last point: I really don't like that the behaviour of ShrinkHelper.exportToServer() is so different to LibertyServer.copyApplicationToServer(). Copying the application to the publish dir seems completely undesirable. I assume it was done to allow people to set up applications in the FATSuite?

This was done because the publish/serverName folder is pushed into the Liberty install upon `LibertyServerFactory.getLibertyServer(name)`, which always occurs before applications are created/exported (typically in the `@BeforeClass` of a test class). So if we only exported to the Liberty install, there would be no record of the application binary anywhere in the autoFVT folder.  This seemed undesirable as I was writing it, because when a failure occurs we zip up the autoFVT folder and that's what we get to debug with, so I figured having the application binaries in there somewhere could be helpful.  This also maintained consistency with the previous behavior where we would package applications at build-time and put them in the publish folder. 

> Ideally I'd prefer exportToServer() to just export to the server and not to publish but I suspect it's too late to change the behaviour of it now. I think a method to just export to the server is necessary (most of the faulttolerance tests should be using it rather than the existing exportToServer()) but I'm not sure what it should be called. Any ideas on that one?

I believe the set of methods you proposed are already covered by `ShrinkHelper.exportArtifact(Archive<?> a, String dest)`.  However, `exportArtifact` is a rather clunky/low-level command because you need to specify the string path to export at.  Perhaps we could also add another boolean to the existing `ShrinkHelper.exportToServer(..., boolean replicate)` methods. I wonder if we're reaching the point where we need a builder pattern though, to avoid all of these overloaded methods?  Something like:
```java
ShrinkBuilder.defaultDropinApp("myApp", "com.foo.web")
  .printArchiveContents(false)
  .replicate(false)
  .export();```

**Comment by Azquelt** [comment]
> I would like to keep forServers() accepting primitive types, since all of this is usually run at static init time. Additionally, this wouldn't work with @Server injection which is undesireable and would confuse most.

Yes, I agree, we would need to make it work with `@Server` injection.

>  I think a good compromise would be passing in String names of servers, and then looking up the servers from within the FeatureReplacementAction

I don't think you can do this. Firstly, `LibertyServer` stores whether or not the server is running as a field in the object. If you get another `LibertyServer` object for the same server, this state will be wrong and you don't know whether or not you need to stop the server.

Secondly, when you call `getLibertyServer()` it actually *deletes* the server directory and re-copies it from the `publish` dir (irrespective of whether or not the server is running, this is the cause of the zombie servers). The result being that it's hard to "look up" a server non-destructively and even if you can, you don't know whether or not it's running.

**exportToServer stuff**

> So if we only exported to the Liberty install, there would be no record of the application binary anywhere in the autoFVT folder. This seemed undesirable as I was writing it, because when a failure occurs we zip up the autoFVT folder and that's what we get to debug with, so I figured having the application binaries in there somewhere could be helpful.

This all makes sense, and we should definitely store the exported apps *somewhere*. The problem as I see it is:
* If you're porting old tests and building your apps in the suite before running any tests, you want to export them to the `publish` directory so that they're copied to the server directory when you do `getLibertyServer()`
* If you're writing new tests and building your app in the test class, you want them to only go to the server.
* If you're building apps in your test classes and put them in the `publish` directory and another test uses the same server, then you're in a world of hurt because
  * The first test app will still be present when the second test runs, which may disrupt the second test
    * Particularly true if the first test checks for validation errors
  *  If you don't run the first test for some reason (maybe you're using skipTests or running in FULL/LITE) then the second test may behave differently because now suddenly the app from the first test *isn't* present
  * If you use `RepeatTests` then it's easy to mess up the test setup. When the test runs the second time, the app is already installed and started by the time you try to export the app again to the server. This can cause problems if you're setting log marks and looking for messages in the logs.

> I believe the set of methods you proposed are already covered [...]

Yes, you can use `exportArtifact` to store it in a file and then `LibertyServer.copyToServerRoot` or whatever to deploy it, but then you have to manually specify paths and stuff and need to be careful to make sure the exported apps end up in the zipped up fat logs.

I believe that exporting directly to the server and storing the app somewhere other than the publish directory should be by far the most common use case for new tests. This should be the simplest thing to do and the path that we promote to test writers.

> ShrinkBuilder

This looks promising and is much clearer than methods with a bunch of boolean flags on them.

Bottom line: at the moment, it's *really hard* to apply `RepeatTests` to a class, due to the way it doesn't handle restarting the server, the way it will wipe the server directory without warning and the behaviour of `exportToServer` described above. Just look at the number of things I had to adjust through trial and error in #2455.

If we think more people will want to do this, we have to make it easier.

**Comment by Azquelt** [comment]
Closing this issue as the problem with the MP and CDI tests is fixed, but we can carry on the discussion about `RepeatTests` and `exportToServer` either here or I can open another issue if you prefer.
